android/app/src/main/java/fi/fmi/mobileweather/BaseWidgetProvider.java

```diff
@@ -513,6 +507,76 @@ else if (background.equals("light")) {
         appWidgetManager.updateAppWidget(appWidgetId, main);
     }
 
+    @Nullable
+    private JSONObject useNewOrStoredJsonObject(JSONObject json, SharedPreferencesHelper pref) {
+        Date now = new Date();
+
+        if (json == null) {
+            // Restore latest json
+
+            long updated = pref.getLong("latest_json_updated", 0L);
+
+            if (updated > (now.getTime() - 24 * 60 * 60 * 1000)) {
+                String jsonstr = pref.getString("latest_json", null);
+
+                try {
+                    json = new JSONObject(jsonstr);
+                } catch (JSONException e) {
+                    showErrorView(
+                            context,
+                            pref,
+                            context.getResources().getString(R.string.update_failed),
+                            context.getResources().getString(R.string.check_internet_connection)
+                    );
+                    return null;
+                }
+            } else {
+                showErrorView(
+                        context,
+                        pref,
+                        context.getResources().getString(R.string.update_failed),
+                        context.getResources().getString(R.string.check_internet_connection)
+                );
+                return null;
+            }
+
+        } else {
+            // Store latest json
+            pref.saveString("latest_json", json.toString());
+            pref.saveLong("latest_json_updated", now.getTime());
+        }
+        return json;
+    }
+
+    @Nullable
+    private JSONArray useNewOrStoredCrisisJsonObject(JSONArray json, SharedPreferencesHelper pref) {
+        Date now = new Date();
+
+        if (json == null) {
+            // Restore latest crisis json
+
+            long updated = pref.getLong("latest_crisis_json_updated", 0L);
+
+            if (updated > (now.getTime() - 24 * 60 * 60 * 1000)) {
```

Comment:
I think that this is too long time for crisis message, because those are not usually valid for so long period. I would use 12 hours initially, but should be discussed before release.

Also would be good if these validity periods (forecast, crisis) would be constants in the beginning of code and therefore easier to change.

Reply:
Fixed

Reply:
Fixed